test/goal_test.exs

```diff
@@ -1025,6 +1041,287 @@ defmodule GoalTest do
+      refute changeset3.valid?
+    end
+
+    test "optional list, min: 1" do
+      schema = %{list: [type: {:array, :string}, min: 1]}
+
+      cs = Goal.build_changeset(schema, %{"list" => []})
+
+      refute cs.valid?
+      assert errors_on(cs) == %{list: ["should have at least 1 item(s)"]}
+    end
+
+    test "list, min: 2" do
+      schema = %{list: [type: {:array, :string}, min: 2]}
+
+      works = %{"list" => ["a", "b"]}
+      breaks = %{"list" => ["a"]}
+      optional = %{}
+
+      changeset1 = Goal.build_changeset(schema, works)
+      changeset2 = Goal.build_changeset(schema, breaks)
+      changeset3 = Goal.build_changeset(schema, optional)
+
+      assert changes_on(changeset1) == %{list: ["a", "b"]}
+      assert changeset1.valid?
+      assert errors_on(changeset2) == %{list: ["should have at least 2 item(s)"]}
+      refute changeset2.valid?
+      assert errors_on(changeset3) == %{}
+      assert changeset3.valid?
+    end
+
+    test "list, max: 2" do
+      schema = %{list: [type: {:array, :string}, max: 2]}
+
+      works = %{"list" => ["a", "b"]}
+      breaks = %{"list" => ["a", "b", "c"]}
+      optional = %{}
+
+      changeset1 = Goal.build_changeset(schema, works)
+      changeset2 = Goal.build_changeset(schema, breaks)
+      changeset3 = Goal.build_changeset(schema, optional)
+
+      assert changes_on(changeset1) == %{list: ["a", "b"]}
+      assert changeset1.valid?
+      assert errors_on(changeset2) == %{list: ["should have at most 2 item(s)"]}
+      refute changeset2.valid?
+      assert errors_on(changeset3) == %{}
+      assert changeset3.valid?
+    end
+
+    test "list, is: 2" do
+      schema = %{list: [type: {:array, :string}, is: 2]}
+
+      works = %{"list" => ["a", "b"]}
+      breaks1 = %{"list" => ["a", "b", "c"]}
+      breaks2 = %{"list" => ["a"]}
+      optional = %{}
+
+      changeset1 = Goal.build_changeset(schema, works)
+      changeset2 = Goal.build_changeset(schema, breaks1)
+      changeset3 = Goal.build_changeset(schema, breaks2)
+      changeset4 = Goal.build_changeset(schema, optional)
+
+      assert changes_on(changeset1) == %{list: ["a", "b"]}
+      assert changeset1.valid?
+      assert errors_on(changeset2) == %{list: ["should have 2 item(s)"]}
+      refute changeset2.valid?
+      assert errors_on(changeset3) == %{list: ["should have 2 item(s)"]}
+      refute changeset3.valid?
+      assert errors_on(changeset4) == %{}
+      assert changeset4.valid?
+    end
+
+    test "list and item both invalid" do
+      schema = %{list: [type: {:array, :string}, max: 1, rules: [min: 3]]}
+
+      works = %{"list" => ["hello"]}
+      breaks = %{"list" => ["as", "we"]}
+
+      changeset1 = Goal.build_changeset(schema, works)
+      changeset2 = Goal.build_changeset(schema, breaks)
+
+      assert changes_on(changeset1) == %{list: ["hello"]}
+      assert changeset1.valid?
+      # todo can/should we only show parent (list) error?
```

Comment:
Let's go with Ecto's default behavior. If `errors_on` returns only the parent or both parent and child errors, let's roll with that 😄